### PR TITLE
feat(iroh-net): Expose DNS resolver

### PR DESCRIPTION
**Pull request by Frando**
## Description

* allow to set a custom DNS resolver in the node builder
* slight refactor of pkarr/dns test utils
* add test using test servers for node discovery to `iroh`

## Change checklist

- [x] Self-review.
- [x] Documentation updates if relevant.
- [x] Tests if relevant.
- [x] All breaking changes documented.

### CODE REVIEW

File: iroh-net/src/discovery.rs [hunk 1]
```diff
@@ -692,24 +684,17 @@ mod test_dns_pkarr {
     }
 
     async fn ep_with_discovery(
-        relay_map: RelayMap,
-        nameserver: SocketAddr,
-        node_origin: &str,
-        pkarr_relay: &Url,
+        relay_map: &RelayMap,
```
Comment (flub) [inline iroh-net/src/discovery.rs:687]: why change this to a reference if it then immediately clones it?  by keeping it as owned it's more obvious to the caller what happens to it and they get to do the clone, if needed, explicitly.

Comment (flub) [reply iroh-net/src/discovery.rs:687]: oh, i guess this is just a test util.  but still, i find that an odd pattern even if it is for tests.  if you really like the convenience of calling it with borrowed data maybe using a generic with `Clone` or `ToOwned` is better.

**Review by flub — APPROVED** [review]

**Review by rklaehn — APPROVED** [review]
very nice!

File: iroh-net/src/discovery.rs (hunk 1 above)
Comment (Frando) [reply iroh-net/src/discovery.rs:687]: I agree in general but thought for a test only util it's ok and makes writing tests a bit simpler IMO. But no strong opinion, can also change back to take by value.

Comment (flub) [reply iroh-net/src/discovery.rs:687]: yeah, i approved anyway as i noticed it's a test.  up to you

**Pull request merged**